File: src/main/java/editor/CustomFontMetrics.java
package editor;

import com.intellij.ide.ui.AntialiasingType;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.impl.EditorImpl;
import com.intellij.openapi.editor.impl.FontInfo;
import com.intellij.util.ui.UIUtil;

import javax.swing.plaf.FontUIResource;
import java.awt.*;
import java.awt.font.FontRenderContext;

class CustomFontMetrics {
    private Editor editor;
    private String familyName;
    private Integer size;
    private FontUIResource font;
    private FontRenderContext context;
    private FontMetrics metrics;
    private Integer lineHeight;

    CustomFontMetrics(Editor editor, String familyName, Integer size) {
        this.editor = editor;
        this.familyName = familyName;
        this.size = size;
        this.font = UIUtil.getFontWithFallback(familyName, Font.PLAIN, size);
        this.context = getCurrentContext(editor);
        this.metrics = FontInfo.getFontMetrics(font, context);
        this.lineHeight = (int) Math.ceil(font.createGlyphVector(context, "Ap").getVisualBounds().getHeight());
    }

    Font getFont() {
        return metrics.getFont();
    }

    Integer getSymbolWidth() {
        return new Canvas().getFontMetrics(font).charWidth('A');
    }

    Boolean isActual(Editor editor, String familyName, Integer size) {
        Font fontActual = metrics.getFont();

        if (!familyName.equals(fontActual.getFamily()) || size != fontActual.getSize()) return false;

        FontRenderContext currentContext = getCurrentContext(editor);
        return currentContext.equals(metrics.getFontRenderContext());
    }

    private FontRenderContext getCurrentContext(Editor editor) {
        FontRenderContext editorContext = FontInfo.getFontRenderContext(editor.getContentComponent());
        Object valueFractionalMetrics = editor.getClass().equals(EditorImpl.class) ? ((EditorImpl) editor).myFractionalMetricsHintValue : RenderingHints.VALUE_FRACTIONALMETRICS_OFF;

        return new FontRenderContext(editorContext.getTransform(), AntialiasingType.getKeyForCurrentScope(false), valueFractionalMetrics);
    }

    Integer lineHeight() {
        return lineHeight;
    }

    FontMetrics metrics() {
        return metrics;
    }
}

File: src/main/scala/editor/CustomFontMetrics.scala
package editor

import java.awt.font.FontRenderContext
import java.awt.{Canvas, Font, FontMetrics, RenderingHints}

import com.intellij.ide.ui.AntialiasingType
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.impl.{EditorImpl, FontInfo}
import com.intellij.util.ui.UIUtil
import javax.swing.plaf.FontUIResource

class CustomFontMetrics(val editor: Editor, val familyName: String, val size: Int) {
  lazy val font: FontUIResource = UIUtil.getFontWithFallback(familyName, Font.PLAIN, size)
  lazy val context: FontRenderContext = getCurrentContext(editor)
  lazy val metrics: FontMetrics = FontInfo.getFontMetrics(font, context)
  lazy val lineHeight: Int = Math.ceil(font.createGlyphVector(context, "Ap").getVisualBounds.getHeight).toInt

  def getFont: Font = metrics.getFont

  def getSymbolWidth: Int = {
    val metrics = new Canvas().getFontMetrics(font)
    metrics.charWidth('A')
  }

  def isActual(editor: Editor, familyName: String, size: Int): Boolean = {
    val font = metrics.getFont
    if (familyName != font.getFamily || size != font.getSize) return false
    val currentContext = getCurrentContext(editor)
    currentContext.equals(metrics.getFontRenderContext)
  }

  private def getCurrentContext(editor: Editor): FontRenderContext = {
    val editorContext = FontInfo.getFontRenderContext(editor.getContentComponent)
    new FontRenderContext(editorContext.getTransform,
      AntialiasingType.getKeyForCurrentScope(false),
      editor match {
        case impl: EditorImpl => impl.myFractionalMetricsHintValue
        case _ => RenderingHints.VALUE_FRACTIONALMETRICS_OFF
      })
  }

}
